File: Collections/NullExpandingDictionary.vb
Namespace Common

    Public Class NullExpandingDictionary

        Public Shared Function Create(Of TKey, TValue)(ByVal Dic As IDictionary(Of TKey, TValue)) As NullExpandingDictionary(Of TKey, TValue)
            Return New NullExpandingDictionary(Of TKey, TValue)(Dic)
        End Function

    End Class

    Public Class NullExpandingDictionary(Of TKey, TValue)
        Inherits BaseDictionary(Of TKey, TValue)

        Public Sub New(ByVal Dic As IDictionary(Of TKey, TValue))
            Me.Dic = Dic
        End Sub

        Public Sub New()
            Me.New(New Dictionary(Of TKey, TValue)())
        End Sub

        Public Overrides ReadOnly Property Count As Integer
            Get
                Return Me.Dic.Count
            End Get
        End Property

        Protected Overrides ReadOnly Property IsReadOnly As Boolean
            Get
                Return Me.Dic.IsReadOnly
            End Get
        End Property

        Default Public Overrides Property Item(key As TKey) As TValue
            Get
                Dim R As TValue
                If Me.Dic.TryGetValue(key, R) Then
                    Return R
                End If
                Return Nothing
            End Get
            Set(value As TValue)
                Me.Dic.Item(key) = value
            End Set
        End Property

        Public Overrides ReadOnly Property Keys As ICollection(Of TKey)
            Get
                Return Me.Dic.Keys
            End Get
        End Property

        Public Overrides ReadOnly Property Values As ICollection(Of TValue)
            Get
                Return Me.Dic.Values
            End Get
        End Property

        Public Overrides Sub Add(key As TKey, value As TValue)
            Me.Dic.Add(key, value)
        End Sub

        Public Overrides Sub Clear()
            Me.Dic.Clear()
        End Sub

        Public Overrides Function ContainsKey(key As TKey) As Boolean
            Return Me.Dic.ContainsKey(key)
        End Function

        Public Overrides Function Remove(key As TKey) As Boolean
            Return Me.Dic.Remove(key)
        End Function

        Public Overrides Function TryGetValue(key As TKey, ByRef value As TValue) As Boolean
            value = Me.Item(key)
            Return True
        End Function

        Protected Overrides Function IEnumerator_1_GetEnumerator() As IEnumerator(Of KeyValuePair(Of TKey, TValue))
            Return Me.GetEnumerator()
        End Function

        Public Function GetEnumerator() As IEnumerator(Of KeyValuePair(Of TKey, TValue))
            Return Me.Dic.GetEnumerator()
        End Function

        Private ReadOnly Dic As IDictionary(Of TKey, TValue)

    End Class

End Namespace

File: Collections/NullExpandingDictionary.cs
using System.Collections.Generic;

namespace Ks.Common
{
    public class NullExpandingDictionary
    {
        public static NullExpandingDictionary<TKey, TValue> Create<TKey, TValue>(IDictionary<TKey, TValue> Dic)
        {
            return new NullExpandingDictionary<TKey, TValue>(Dic);
        }
    }

    public class NullExpandingDictionary<TKey, TValue> : BaseDictionary<TKey, TValue>
    {
        public NullExpandingDictionary(IDictionary<TKey, TValue> Dic)
        {
            this.Dic = Dic;
        }

        public NullExpandingDictionary() : this(new Dictionary<TKey, TValue>())
        {
        }

        public override int Count => this.Dic.Count;

        protected override bool IsReadOnly => this.Dic.IsReadOnly;

        public override TValue this[TKey key]
        {
            get
            {
                if (this.Dic.TryGetValue(key, out var R))
                {
                    return R;
                }

                return default;
            }
            set => this.Dic[key] = value;
        }

        public override ICollection<TKey> Keys => this.Dic.Keys;

        public override ICollection<TValue> Values => this.Dic.Values;

        public override void Add(TKey key, TValue value)
        {
            this.Dic.Add(key, value);
        }

        public override void Clear()
        {
            this.Dic.Clear();
        }

        public override bool ContainsKey(TKey key)
        {
            return this.Dic.ContainsKey(key);
        }

        public override bool Remove(TKey key)
        {
            return this.Dic.Remove(key);
        }

        public override bool TryGetValue(TKey key, out TValue value)
        {
            value = this[key];
            return true;
        }

        protected override IEnumerator<KeyValuePair<TKey, TValue>> _GetEnumerator()
        {
            return this.GetEnumerator();
        }

        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
        {
            return this.Dic.GetEnumerator();
        }

        private readonly IDictionary<TKey, TValue> Dic;
    }
}
